### PR TITLE
Refactor Package into ROS1 and ROS2 versions

### PR DESCRIPTION
**Pull request by schmerl**
(Note currently ROS2Package and ROS1Package are identical and will be changed in subsequent tasks of #420)

### CODE REVIEW

File: src/roswire/common/package.py [hunk 1]
```diff
@@ -116,116 +69,19 @@ class PackageDatabase(ABC, Mapping[str, Package]):
...
 
-    @classmethod
-    def build(
-        cls, app_instance: "AppInstance", paths: Optional[List[str]] = None
-    ) -> "PackageDatabase":
-        if paths is None:
-            paths = cls._determine_paths(app_instance)
-        db_package = cls._from_paths(app_instance, paths)
-        return db_package
+    _paths_in_package: List[str]
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:72]: If this is intended as a protected attribute, it should move to the same location as the other attributes. Also, does the type annotation need to be `List`? Or is `Sequence` or `Collection` sufficient?

File: src/roswire/common/package.py [hunk 2]
```diff
@@ -116,116 +69,19 @@ class PackageDatabase(ABC, Mapping[str, Package]):
...
-
-        Parameters
-        ----------
-        app_instance: AppInstance
-            An instance of the application for which the
-            list of paths should be obtained
-        """
+    def build(cls,
+              app_instance: "AppInstance",
+              paths: Optional[List[str]]) -> "PackageDatabase":
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:78]: Nitpick: return type should be on a separate line.
More importantly, `paths` should have a default value of `None`.

File: src/roswire/common/package.py [hunk 3]
```diff
@@ -116,116 +69,19 @@ class PackageDatabase(ABC, Mapping[str, Package]):
...
-                packages.append(package)
-        return cls._from_packages_and_paths(packages, paths)
-
-    def __init__(
-        self, packages: Iterable[Package], paths: Iterable[str]
-    ) -> None:
-        self.__contents = {p.name: p for p in packages}
-        self._paths_in_package = list(paths)
-
     @property
```
Comment (ChrisTimperley) [inline src/roswire/common/package.py:81]: If we have `paths` as an immutable attribute, we won't need this getter.

File: src/roswire/ros1/package.py [hunk 4]
```diff
@@ -3,27 +3,160 @@
...
-        cls, packages: Iterable[Package], paths: Iterable[str]
-    ) -> "PackageDatabase":
+            cls, packages: Iterable[ROS1Package], paths: Iterable[str]
+    ) -> "ROS1PackageDatabase":
         return ROS1PackageDatabase(packages, paths)
 
     @classmethod
-    def from_dict(cls, d: List[Dict[str, Any]]) -> "PackageDatabase":
-        return ROS1PackageDatabase([Package.from_dict(dd) for dd in d], [])
+    def from_dict(cls, d: List[Dict[str, Any]]) -> "ROS1PackageDatabase":
```
Comment (ChrisTimperley) [inline src/roswire/ros1/package.py:104]: Return type should be `ROSPackageDatabase` to avoid breaking LSP.

File: src/roswire/ros1/package.py [hunk 5]
```diff
@@ -3,27 +3,160 @@
...
+            "messages": [m.to_dict() for m in self.messages],
+            "services": [s.to_dict() for s in self.services],
+            "actions": [a.to_dict() for a in self.actions],
+        }
+        return d
+
+
 class ROS1PackageDatabase(PackageDatabase):
+
+    __contents: Dict[str, Any]
```
Comment (ChrisTimperley) [inline src/roswire/ros1/package.py:95]: I'm not entirely sure what this is.

File: src/roswire/common/package.py (hunk 1 above)
Comment (schmerl) [reply src/roswire/common/package.py:72]: The property paths() accesses this, so I don't think I can move it. (Or I need to duplicate it down in all subclasses.)

File: src/roswire/ros2/package.py [hunk 6]
```diff
@@ -50,3 +179,30 @@ def _determine_paths(cls, app_instance: "AppInstance") -> List[str]:
             for (package, prefix) in package_to_prefix.items()
         ]
         return paths
+
+    def __init__(
+            self, packages: Iterable[ROS2Package], paths: Iterable[str]
```
Comment (ChrisTimperley) [inline src/roswire/ros2/package.py:184]: Excessive indentation. Readability could be improved by using a separate line for each argument.

File: src/roswire/common/package.py (hunk 2 above)
Comment (schmerl) [reply src/roswire/common/package.py:78]: Done

File: src/roswire/ros1/package.py [hunk 7]
```diff
@@ -1,29 +1,110 @@
...
+@attr.s(frozen=True, auto_attribs=True, slots=True)
+class ROS1Package(Package[MsgFormat, SrvFormat, ActionFormat]):
+    name: str
+    path: str
+    messages: Collection[MsgFormat] = attr.ib(converter=tuple_from_iterable)
+    services: Collection[SrvFormat] = attr.ib(converter=tuple_from_iterable)
+    actions: Collection[ActionFormat] = attr.ib(converter=tuple_from_iterable)
+
+    @classmethod
+    def build(cls, path: str, app_instance: "AppInstance") -> "ROS1Package":
```
Comment (ChrisTimperley) [inline src/roswire/ros1/package.py:30]: This violates LSP. (Return type of `build` in `Package` is `Package`.) We'll need to think about how to deal with this one.

Comment (ChrisTimperley) [reply src/roswire/ros1/package.py:30]: One way to avoid this issue is to move this logic into the `_build_package` method of `PackageDatabase`.

**Review by ChrisTimperley — APPROVED** [review]
👍🏻 Awesome. Let's get it merged!

**Pull request merged**